In [1]:
import pandas as pd

In [2]:
data_dir = "./data/raw_data/"
save_dir = "./data/preprocessed_data/"

Read in the data

In [3]:
cv_table = pd.read_csv(data_dir + "cv_table.csv")
ndr_table = pd.read_csv(data_dir + "ndr_table.csv")
poll_table = pd.read_csv(data_dir + "poll_table.csv")

latlong_cv = pd.read_csv(data_dir + "latlong_cv.csv")
latlong_ndr = pd.read_csv(data_dir + "latlong_ndr.csv")
latlong_pollination = pd.read_csv(data_dir + "latlong_pollination.csv")


We have three tables of data, one for each contribution of nature. We also have three tables relating each data point to a longitude and a latitude.

Let's look at the tables first!

# cv_table

In [4]:
cv_table.describe(include = 'all')

fid         UN_cur        UN_ssp1        UN_ssp3  \
count  686665.000000  686665.000000  686665.000000  686665.000000   
mean   343333.000000       2.482690       2.822749       2.982682   
std    198223.255633       0.511879       0.573237       0.618028   
min         1.000000       1.030057       1.237034       1.297790   
25%    171667.000000       2.082644       2.376177       2.492883   
50%    343333.000000       2.440013       2.777381       2.935599   
75%    514999.000000       2.817269       3.259844       3.419952   
max    686665.000000       4.291871       4.817462       5.000000   

             UN_ssp5        pop_cur       pop_ssp1       pop_ssp3  \
count  686665.000000  686665.000000  686665.000000  686665.000000   
mean        3.139066     485.270589     586.541642     605.606714   
std         0.644773    2772.848940    3590.280849    3368.830041   
min         1.311983       0.000000       0.000000       0.000000   
25%         2.667168       0.510938       0.423553       0.455772   
50%         3.107233      21.314615      21.012911      26.209813   
75%         3.619904     156.068106     153.478002     184.031578   
max         5.000000  236893.117288  270175.660403  247236.351206   

            pop_ssp5  
count  686665.000000  
mean      608.113528  
std      3719.613422  
min         0.000000  
25%         0.445830  
50%        21.459414  
75%       159.716874  
max    270048.141080

This table describes the coastal risk mitigation by natural habitats.
Some explanation of the table follow:

* **fid** - the id of the geographical location that the data describes. This id can be converted to lon/lat using the latlong table.  
* **UN_cur** - the **current** (2015) unmet need of the coastal risk mitigation in this particular area. If this figure is negative then it means that we have enough, and even a surplus, in this particular area. 
* **UN_ssp1** the unmet need of the coastal risk mitigation in the year 2050, given that scenario **ssp1** is realized. 
* **ssp1, ssp3, ssp5** - the three future scenarios. It is not perfectly clear which scenraio is which here. ssp1 is most likely the sustainability scenario, purely based on the observation that the unmet need in this column is lower than for the other scenarios. The other two scenarios are fossil-fuels and regional rivalry.
* **pop_cur** - the **current** (2015) number of people in this geographic region who are affected by coastal risk mitigation. If the unmet need is high and the population is high then a lot of people are negatively affected.
* **pop_sspX** - same as the above, but in the year 2050, for the three different future scenarios.

# ndr_table

In [5]:
ndr_table.describe(include = 'all')

fid     country   region        UN_cur       UN_ssp1  \
count   25243.000000       25243    25243  2.524300e+04  2.524300e+04   
unique           NaN         228        8           NaN           NaN   
top              NaN  Antarctica  UNKNOWN           NaN           NaN   
freq             NaN        6511     6557           NaN           NaN   
mean    31136.198669         NaN      NaN  2.847848e+06  2.982161e+06   
std     20508.712007         NaN      NaN  8.093390e+06  9.765498e+06   
min      2262.000000         NaN      NaN  0.000000e+00  0.000000e+00   
25%     13083.500000         NaN      NaN  0.000000e+00  0.000000e+00   
50%     25299.000000         NaN      NaN  2.304164e+03  7.343517e+02   
75%     56810.500000         NaN      NaN  2.418531e+06  1.757585e+06   
max     64799.000000         NaN      NaN  1.714816e+08  3.159408e+08   

             UN_ssp3       UN_ssp5     rurpopcur    rurpopssp1    rurpopssp3  \
count   2.524300e+04  2.524300e+04  2.524300e+04  2.524300e+04  2.524300e+04   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean    3.464451e+06  2.888546e+06  1.693007e+05  1.008505e+05  2.328594e+05   
std     1.072992e+07  8.570622e+06  6.919647e+05  4.071316e+05  9.817925e+05   
min     0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%     0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%     7.217091e+02  7.175778e+02  0.000000e+00  0.000000e+00  0.000000e+00   
75%     2.092982e+06  1.920021e+06  4.479987e+04  2.482487e+04  4.230556e+04   
max     2.093042e+08  2.064157e+08  1.667823e+07  8.933015e+06  2.162865e+07   

          rurpopssp5  
count   2.524300e+04  
unique           NaN  
top              NaN  
freq             NaN  
mean    1.003789e+05  
std     4.030947e+05  
min     0.000000e+00  
25%     0.000000e+00  
50%     0.000000e+00  
75%     2.592468e+04  
max     8.607843e+06

This table describes the improvement of water quality by filtering of pollutants (in our case probably excessive amounts of nutrients, making the water quality bad) when water passes by natural habitats such as forests and wetlands.

The columns are the same as in the previous table, although `pop` is replaced by `rurpop`. Charlotte has confirmed that this is still the same metric though.

We also have two extra columns `country` and `region`. We will most probably not use these, especially as we don't actually have them for each table.

One thing we can immediately observe is that the count (number of entries) is much lower, the previous table had 686665 entries while this one has only 25243. This needs to be investigated. Does this mean that we don't have data for the whole earth for this particular contribution of nature?

# poll_table

In [6]:
poll_table.describe(include = 'all')

fid        UN_cur       UN_ssp1       UN_ssp3       UN_ssp5  \
count   9160.000000  9.160000e+03  9.160000e+03  9.160000e+03  9.160000e+03   
mean   25270.656441  9.630618e+04  8.972863e+04  9.375806e+04  9.990251e+04   
std    10032.053267  4.348216e+05  4.136170e+05  4.113846e+05  4.335244e+05   
min     8845.000000  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    16672.000000  2.088596e+00  2.046068e+00  9.315915e+00  9.550000e+00   
50%    22971.500000  1.027879e+03  9.781146e+02  1.492830e+03  1.594218e+03   
75%    33446.250000  3.107308e+04  2.827017e+04  3.439839e+04  3.474784e+04   
max    51951.000000  2.322062e+07  2.249534e+07  2.321814e+07  2.270223e+07   

            pop_cur      pop_ssp1      pop_ssp3      pop_ssp5  
count  9.160000e+03  9.160000e+03  9.160000e+03  9.160000e+03  
mean   2.574799e+05  3.993871e+05  4.659653e+05  3.603159e+05  
std    1.004224e+06  1.546353e+06  1.681148e+06  1.450151e+06  
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
50%    6.223414e+02  3.940410e+03  3.460522e+03  0.000000e+00  
75%    1.263994e+05  1.874406e+05  2.276243e+05  1.468894e+05  
max    2.338777e+07  3.778232e+07  3.282686e+07  3.222564e+07

This table describes the pollination of crops coming from natural habitats inhabiting pollinators (ex. bees).

The columns are the same as previously.

Here though we have an even lower count. What does this mean? A hypothesis could be that the resolution is simply not just not as high for the data here.

# latlong_cv

In [7]:
latlong_cv.describe(include = 'all')

fid            lat            lng
count  686665.000000  686665.000000  686665.000000
mean   343333.000000       9.346826      23.706301
std    198223.255633      32.202816     102.750279
min         1.000000     -58.480000    -180.000000
25%    171667.000000     -10.600000     -73.070000
50%    343333.000000       9.600000      12.640000
75%    514999.000000      35.450000     123.250000
max    686665.000000      62.800000     180.000000

This looks good, just a simple mapping from the **fid** to the coordinates. We have the same number of mappings as we have entrys in the `cv_table`.

# latlong_ndr

In [8]:
latlong_ndr.describe(include = 'all')

fid           lat           lng
count  64800.000000  64800.000000  64800.000000
mean   32399.500000      0.000000      0.000000
std    18706.293059     51.961123    103.923449
min        0.000000    -89.500000   -179.500000
25%    16199.750000    -44.750000    -89.750000
50%    32399.500000      0.000000      0.000000
75%    48599.250000     44.750000     89.750000
max    64799.000000     89.500000    179.500000

Here the number of entries is actually higher than the number of entries in the `ndr_table`. This is a bit strange, what does this mean?

Moreover, why do we even need to have different mappings for the different contributions of nature? Couldn't they just use the same mappings between id and coordinates? The answer to this is probably that, as suggested before, the world map resolutions of the nature's contributions are different, with coastal risk having the most detailed data.

# latlong_pollination

In [9]:
latlong_pollination.describe(include = 'all')

fid           lat           lng
count  64800.000000  64800.000000  64800.000000
mean   32399.500000      0.000000      0.000000
std    18706.293059     51.961123    103.923449
min        0.000000    -89.500000   -179.500000
25%    16199.750000    -44.750000    -89.750000
50%    32399.500000      0.000000      0.000000
75%    48599.250000     44.750000     89.750000
max    64799.000000     89.500000    179.500000

Once again the same thing, the number of entries are the same as in the `latlong_ndr` table, 64800, while the number of entries in the `poll_table` is just 9160...



# Actual data

We also take a look at the actual data, just to get a quick feel for the actual figures:

In [10]:
cv_table.head()

fid    UN_cur   UN_ssp1   UN_ssp3   UN_ssp5  pop_cur  pop_ssp1  pop_ssp3  \
0    1  3.282099  3.684031  3.684031  4.135186      0.0       0.0       0.0   
1    2  2.924018  3.282099  3.282099  3.684031      0.0       0.0       0.0   
2    3  2.924018  3.282099  3.282099  3.684031      0.0       0.0       0.0   
3    4  2.924018  3.282099  3.282099  3.684031      0.0       0.0       0.0   
4    5  2.924018  3.282099  3.282099  3.684031      0.0       0.0       0.0   

   pop_ssp5  
0       0.0  
1       0.0  
2       0.0  
3       0.0  
4       0.0

Nothing surprising here! The id:s start from 1 and go up. We see an unmet need for the first 5 geographical regions both currently and for the 3 different future scenarios. We also see that nobody in this region actually gets directly affected by any changes to the natural habitats mitigating coastal risk. This could be because nobody lives there, or simply perhaps because there is no coast in this area.

Let's look at the `ndr_table`.

In [11]:
ndr_table.sample(5)

fid                           country   region        UN_cur  \
16104  36205  Democratic Republic of the Congo   Africa  1.674984e+06   
6441   13262                            Russia  Eurasia  4.045899e+05   
15470  34174                  Papua New Guinea  Oceania  0.000000e+00   
23819  63376                        Antarctica  UNKNOWN  0.000000e+00   
14194  30058                             Ghana   Africa  1.167166e+06   

            UN_ssp1       UN_ssp3       UN_ssp5     rurpopcur     rurpopssp1  \
16104  6.911775e+05  8.115862e+05  8.144077e+05  2.648934e+05  230902.578100   
6441   4.012163e+05  4.152183e+05  3.931995e+05  1.185113e+04    7582.272461   
15470  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00       0.000000   
23819  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00       0.000000   
14194  1.883950e+06  1.306541e+06  4.207906e+06  1.268533e+06   97216.562500   

         rurpopssp3     rurpopssp5  
16104  7.000974e+05  230909.906300  
6441   9.816912e+03    8103.152832  
15470  0.000000e+00       0.000000  
23819  0.000000e+00       0.000000  
14194  1.231804e+06   95833.234380

We note that the **fid** goes above the maximum number of entries in the table (which was 25243). 

Same for `poll_table`:

In [12]:
poll_table.sample(5)

fid       UN_cur      UN_ssp1      UN_ssp3      UN_ssp5       pop_cur  \
6668  32621  4168.757878  4257.722532  4393.030225  4392.279882   93224.70313   
7174  34691    47.156799    45.831140    53.051223    93.809180  234947.28130   
7706  37185   171.682375   165.798171   288.975228   183.196205   92237.45313   
3446  19340    81.112525    67.935944    50.390224    81.112525   30059.59766   
7075  34322     0.000000     0.000000     0.000000     0.000000    7878.15332   

           pop_ssp1       pop_ssp3       pop_ssp5  
6668  133665.031300  234833.500000  114015.726600  
7174  255341.640600  304156.812500  253622.156300  
7706       0.000000  125735.500000       0.000000  
3446   22724.455080   27804.460940   22570.267580  
7075    7933.297363    9833.271484    7882.144531

And for an example lat_long-table let's check out `latlong_cv`:

In [13]:
latlong_cv.sample(5)

fid    lat     lng
363092  363093  10.24  -64.58
466735  466736  25.03 -108.04
538601  538602  39.10  141.90
568288  568289  47.63  -59.26
341285  341286   9.28   80.79

# Preprocessing

The next step is to perform the preprocessing (if needed) and save the data in new files. 
This might need to be done in several iteration as we don't know at the moment exactly how we will use the data. One thing that will be good to have done already is to join the lat/lon tables with the data tables.

Start with `cv_table`:

In [14]:
cv_table_out = pd.merge(cv_table, latlong_cv, on='fid')
cv_table_out.sample()

fid   UN_cur   UN_ssp1   UN_ssp3   UN_ssp5   pop_cur  pop_ssp1  \
34493  34494  2.58734  2.904191  2.904191  3.259844  0.013091       0.0   

       pop_ssp3  pop_ssp5    lat    lng  
34493       0.0       0.0 -48.63 -75.58

`fid` column is not needed, drop this.

In [15]:
cv_table_out.drop(columns=['fid'], inplace=True)
cv_table_out.sample()

UN_cur   UN_ssp1   UN_ssp3   UN_ssp5  pop_cur  pop_ssp1  pop_ssp3  \
671814  2.685377  3.014233  3.224968  3.224968      0.0       0.0       0.0   

        pop_ssp5    lat    lng  
671814       0.0  60.86 -47.93

Looks good! Let's do the same for `ndr_table` and `poll_table`:

In [16]:
ndr_table_out = pd.merge(ndr_table, latlong_ndr, on ='fid')

# Remove the country and region columns, which we will probably not use, and rename columns
ndr_table_out.drop(columns=['fid', 'country', 'region'], inplace=True) 
ndr_table_out.rename(columns={"rurpopcur": "pop_cur", 
                              "rurpopssp1": "pop_ssp1", 
                              "rurpopssp3": "pop_ssp3",
                              "rurpopssp5": "pop_ssp5"}, inplace=True)

ndr_table_out.sample()

UN_cur  UN_ssp1  UN_ssp3  UN_ssp5  pop_cur  pop_ssp1  pop_ssp3  \
16603     0.0      0.0      0.0      0.0      0.0       0.0       0.0   

       pop_ssp5   lat    lng  
16603       0.0 -16.5 -150.5

In [17]:

poll_table_out = pd.merge(poll_table, latlong_pollination, on='fid')
poll_table_out.drop(columns=['fid'], inplace=True)

poll_table_out.sample()

UN_cur      UN_ssp1      UN_ssp3      UN_ssp5    pop_cur  \
3082  11205.57207  10341.60583  11426.92235  11698.08962  1213828.0   

        pop_ssp1    pop_ssp3    pop_ssp5   lat   lng  
3082  1370219.75  1067409.25  1506498.25  38.5  15.5

Now we are ready to save the files!

# Saving data

In [20]:
cv_table_out.to_csv(save_dir + "cv_table_preprocessed.csv.gz", compression='gzip')
ndr_table_out.to_csv(save_dir + "ndr_table_preprocessed.csv.gz", compression='gzip')
poll_table_out.to_csv(save_dir + "poll_table_preprocessed.csv.gz", compression='gzip')